cpp/src/text/subword/wordpiece_tokenizer.cu

```diff
@@ -0,0 +1,442 @@
+                                                       uint8_t* tokens_per_word)
+{
+  uint32_t char_for_thread = blockDim.x * blockIdx.x + threadIdx.x;
+
+  // Deal with the start_word_indices array
+  if (char_for_thread < num_code_points) {
+    uint32_t val_to_write = std::numeric_limits<uint32_t>::max();
+    if ((code_points[char_for_thread] != SPACE_CODE_POINT) && (char_for_thread > 0) &&
+        (code_points[char_for_thread - 1] == SPACE_CODE_POINT)) {
+      val_to_write = char_for_thread;
+    }
+    start_word_indices[char_for_thread] = val_to_write;
+
+    // Deal with the end_word_indices_array
+    val_to_write = std::numeric_limits<uint32_t>::max();
+    if ((code_points[char_for_thread] != SPACE_CODE_POINT) &&
+        (char_for_thread + 1 < num_code_points) &&
+        (code_points[char_for_thread + 1] == SPACE_CODE_POINT)) {
+      val_to_write = char_for_thread + 1;
+    }
+    end_word_indices[char_for_thread] = val_to_write;
+
+    token_ids[char_for_thread]       = std::numeric_limits<uint32_t>::max();
+    tokens_per_word[char_for_thread] = 0;
+  }
+}
+
+/**
+ * @brief Resolves the string boundaries for the start and end words.
+ *
+ * This kernel should be called after `mark_word_start_and_ends` with at
+ * least `num_strings` total threads.
+ *
+ * The start and end indices are updated to honor the string boundaries
+ * within the strings array. This corrects any word ranges that span across
+ * individual strings.
+ *
+ * @param code_points A pointer to the code points in the strings.
+ * @param strings_offsets An array containing the index of the starting character of each string
+ *        with an extra space at the end containing the total number of characters. As a result,
+ *        this array is of length num_strings + 1.
+ * @param start_word_indices An array which will contain the starting index for each word scattered
+ *        throughout. If an index does not represent a word start, the max-uint32_t value is written
+ *        to indicate this.
+ * @param end_word_indices An array which will contain the one past the end index for each word
+ *        scattered throughout. If an index does not represent a word end, the max uint32_t value is
+ *        written to indicate this.
+ * @param num_strings The total number of strings to be processed.
+ */
+__global__ void mark_string_start_and_ends(uint32_t const* code_points,
+                                           uint32_t const* strings_offsets,
+                                           uint32_t* start_word_indices,
+                                           uint32_t* end_word_indices,
+                                           uint32_t num_strings)
+{
+  uint32_t idx = blockDim.x * blockIdx.x + threadIdx.x;
+  // Ensure the starting character of each strings is written to the word start array.
+  if (idx <= num_strings) {
+    auto const offset = strings_offsets[idx];
+
+    if ((idx < num_strings) && (code_points[offset] != SPACE_CODE_POINT)) {
```

Comment:
Do you really need both start and end indices? Isn't the start of word n+1 just the end of word n, plus 1? In which case this just becomes a thrust::transform with a lambda.

Reply:
The start of word `n+1` is not the end of word `n` since there can be more than one whitespace character delimiting the tokens (after normalizing). Regardless, the start and end indices never intersect and so I think there is room to improve this. But I feel the amount of work here would be out of scope for this port and should be done in a follow on PR.